File: solutions/clojure/scrabble-score/1/src/scrabble_score.clj
(ns scrabble-score)

(def reverse-scores
  {1 "AEIOULNRST"
   2 "DG"
   3 "BCMP"
   4 "FHVWY"
   5 "K"
   8 "JX"
   10 "QZ"})

(def scores
  (->> (for [[score ltrs] reverse-scores l ltrs] [l score])
       (into {})))


(def score-letter (comp scores first clojure.string/upper-case))

(defn score-word [word]
  (->> word
       (map score-letter)
       (apply +)))

File: solutions/scala/scrabble-score/1/src/main/scala/ScrabbleScore.scala
object ScrabbleScore {
  private val letterScores =
    Map(
      "aeioulnrst" -> 1,
      "dg" -> 2,
      "bcmp" -> 3,
      "fhvwy" -> 4,
      "k" -> 5,
      "jx" -> 8,
      "qz" -> 10
    ).flatMap { case (letters, score) => letters.map(_ -> score) }

  def score(word: String): Int =
    word.toLowerCase.map(letterScores).sum
}
